pallets/asset/src/lib.rs

```diff
@@ -2176,6 +2197,35 @@ impl<T: Trait> Module<T> {
         (transfer_status, weight_for_valid_transfer)
     }
 
+    /// Sets the initial total supply of a confidential asset.
+    /// Only called by the token owner of a confidential asset. Can be called only once.
+    ///
+    /// # Arguments
+    /// * `origin` Secondary key of the token owner.
+    /// * `ticker` Ticker of the token.
+    /// * `total_supply` Ticker of the token.
+    pub fn unchecked_set_total_supply(
+        did: IdentityId,
+        ticker: Ticker,
+        total_supply: T::Balance,
+    ) -> DispatchResult {
```

Comment:
I don't think we need `is_owner()` check here as it is already checked in the `mint_confidential_asset()` dispatch. similarly, we don't need 

```
ensure!(
            total_supply != Zero::zero(),
            Error::<T>::TotalSupplyMustBePositive
        );
```
this check should be part of the preliminary sanity checks. 

Even any of the checks aren't useful here because you are already mutating the storage before these checks in the confidential_asset pallet so even if they fail storage changes will remain to persist. 

Writing dispatchable in the substrate has a thumb rule - Verify first, Write last.

So IMO you only need a functionality here that changes storage without any checks (that justify function name `unchecked` 😉 ). and every check can be performed at the dispatchable level.